### PR TITLE
[MAINTENANCE] Optimize CategoricalColumnDomainBuilder to compute metrics in a single method call

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GREAT-464/GREAT-498/GREAT-638
-
-

After submitting your PR, CI checks will run and @ge-cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def name(self) -> str:
         return self._name
 
     @property
-    def json_serialize(self) -> bool:
+    def json_serialize(self) -> Union[str, bool]:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_builder.py:176]: 🙇🏽

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py [hunk 2]
```diff
@@ -182,67 +182,100 @@ def _generate_metric_configurations_to_check_cardinality(
...
+
+        # Step 6: Produce "column_name" list, corresponding to effective "MetricConfiguration" ID values.
+        candidate_column_names: List[str] = [
+            column_name
+            for column_name, metric_configuration_ids in metric_configuration_ids_by_column_name.items()
+            if all(
+                [
+                    metric_configuration_id in metric_configuration_ids_resolved_metrics
+                    for metric_configuration_id in metric_configuration_ids
+                ]
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:279]: Nitpick - these can be generators so we don't have to create all the objects

**Review by cdkini — APPROVED** [review]
LGTM

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:279]: @cdkini Is it a quick fix?  If so, could you please suggest the pattern you had in mind?  Thanks!

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py [hunk 3]
```diff
@@ -166,7 +166,7 @@ def _generate_metric_configurations_to_check_cardinality(
         self,
         batch_ids: List[str],
         column_names: List[str],
-    ) -> List[Dict[str, List[MetricConfiguration]]]:
+    ) -> Dict[str, List[MetricConfiguration]]:
```
Comment (anthonyburdi) [inline great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:169]: Please also change the docstring to match the new return type

**Review by anthonyburdi — APPROVED** [review]
Just a small request but will approve to not block

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:169]: @anthonyburdi Done (in next commit).  Thanks!

**Pull request merged**